Enables spillable/unspillable state for RapidsBuffer and allow buffer sharing

abellina opened the pull request:
Signed-off-by: Alessandro Bellina <abellina@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/6864
Closes https://github.com/NVIDIA/spark-rapids/issues/6561

This PR allows a RapidsBuffer to become "non spillable" for the period of time that a `ColumnarBatch` is obtained from it (and the ref count of the device memory buffer is > 1). It also enables sharing RapidsBuffer between different RapidsBufferHandle instances. This means that when DeviceMemoryBuffers are being added to the spill framework, we detect whether we are already tracking that buffer (defined as between the creation of the `RapidsDeviceMemoryBuffer` and the first time `free` is called on it).

This is a draft PR since I am still running some performance numbers and I need to add more unit tests. There are various race conditions that this creates. Because of the race conditions we may put this into 23.04 instead.

These are tests I need to document, some of them I have done:

- [x] Performance testing with NDS @ 3TB
- [x] Stress testing with shared RapidsBuffers at 3TB
- [x] Stress testing with query95 3TB, 10TB, 30TB 
- [ ] Add more unit tests to exercise potential race conditions
- [x] Verify with nsys that extra calls to `Table.columnViewsFromPacked` are as cheap as we think

revans2 reviewed (commented):
I did a first pass through this and it looks good, but the code is complicated enough I need to dig in much deeper.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala:
```diff
@@ -144,6 +213,14 @@ abstract class RapidsBufferStore(
+  }
+
+  protected def setSpillable(buffer: RapidsBufferBase, isSpillable: Boolean): Unit = {
+    throw new NotImplementedError(s"This store ${this} does not implement markSpillable")
```
```suggestion
    throw new NotImplementedError(s"This store ${this} does not implement setSpillable")
```

abellina commented:
build

abellina commented:
build

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:
```diff
@@ -256,35 +267,88 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
     override val storageTier: StorageTier = StorageTier.DEVICE
 
-    override protected def releaseResources(): Unit = {
+    require(contigBuffer.setEventHandler(this) == null,
```
If this does throw an exception how bad of a state is the spill code? I mean we just changed who the event handler is pointing to, should we exit? should it be a fatal error? should we not change the event handler unless it is exactly what we expect? I am just a little concerned about how we might recover from this.

abellina replied:
I have changed the call instantiating `RapidsDeviceMemoryBuffer` to hold the `contigBuffer` lock. Before instantiating this we go through the event handler detection logic as well, so this is the new order of operation:

1. Lock the cuDF memory buffer
2. Try to get a handler from the memory buffer. If not set, we will create a new `RapidsDeviceMemoryBuffer`. If set, we call `addReference` on the `RapidsDeviceMemoryBuffer` and if that succeeds, we know we were valid and are _at worst_ in the pending queue to be spilled. If we can't add a reference we lost the race, and the buffer has since spilled to host, or has been removed.
3. We instantiate a new `RapidsDeviceMemoryBuffer` if (2) failed to find a handler, or failed to `addReference`.
4. We associate this new `RapidsDeviceMemoryBuffer` with the handler.
5. Unlock cuDF memory buffer

I do not think it's possible we wind up here. I'd like this to be a loud exception since it would be a very bad thing if we overwrote that handler. So I expect no recovery from this scenario. @revans2 please let me know if the logic makes sense.

revans2 reviewed (commented):
Mostly just some nits. Feeling better about this change.

abellina commented:
I am currently running stress testing for NDS q72 at 3TB by setting the percent of GPU allocation to 30% of the original and watching it spill, especially spill broadcast (shared rapids buffers). I am debugging an inc after close seen here.

abellina commented:
> I am debugging an inc after close seen here

The INC AFTER CLOSE issue is not related to my change. Some of the gather code has race conditions when an exception is thrown (OOM in this case) and can cause this. It is more of a nuisance than anything as the executor is going down anyway. Filed this to look at separately: https://github.com/NVIDIA/spark-rapids/issues/7581

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:
```diff
@@ -51,49 +51,46 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
-      addDeviceBuffer(buffer, needsSync = true)
-      catalog.makeNewHandle(id, initialSpillPriority, spillCallback)
+  def getExistingRapidsBufferAndAcquire(
+      buffer: DeviceMemoryBuffer): Option[RapidsDeviceMemoryBuffer] = {
```
Do we need synchronization here to solve the case of multiple threads all referencing the same buffer and all deciding to make it spillable at the same time?

abellina replied:
Yeah this is similar to @revans2 comment. It was missed. 

It would be bad if 2 callers want to add the same buffer and they are added as separate RapidsDeviceMemoryBuffers. It would trigger a require I put into this class if the event handler was already set at runtime and that was not intended.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala:
```diff
@@ -76,6 +132,7 @@ abstract class RapidsBufferStore(
         val buffs = buffers.values().toArray(new Array[RapidsBufferBase](0))
         buffers.clear()
         spillable.clear()
+        spilling.clear()
```
right below this we call `safeFree` on all the buffers, which updates the counters, so this is handled but via each buffer getting freed.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:
```diff
@@ -119,12 +116,20 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
+        catalog.makeNewHandle(id, initialSpillPriority, spillCallback)
+      }
+    }.getOrElse {
+      addContiguousTable(
```
I have changed the code to lock the underlying buffer for this both `add*` methods, so we should be covered.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/DeviceMemoryEventHandler.scala:
```diff
@@ -104,32 +104,37 @@ class DeviceMemoryEventHandler(
    * @return true if allocation should be reattempted or false if it should fail
    */
-  override def onAllocFailure(allocSize: Long, retryCount: Int): Boolean = {
+  override def onAllocFailure(allocSize: Long, retryCount: Int): Boolean = synchronized {
```
I have rolled back the synchronize change here. I am going to add the atomic integer soon, just need to test it some.

abellina replied:
@jlowe added https://github.com/NVIDIA/spark-rapids/pull/7572/commits/51673a18756bce2269d7f265a15da8afe216c5c2 to handle this scenario

abellina commented:
In my stress testing at 3TB I found a deadlock due to lock inversion with a new commit I added given the race condition feedback (https://github.com/NVIDIA/spark-rapids/pull/7572/commits/8737302d24d87ae7fbd313a5af985b771c0fbe92). I am debugging it.

abellina commented:
Performance-wise, overall for NDS 3TB with 5 repetitions for baseline (I used a 23.02 snapshot from last week) vs this change:
```
Name = benchmark
Means = 533737.6, 533441.6
Time diff = 296.0
Speedup = 1.0005548873578665
T-Test (test statistic, p value, df) = 0.09851820421116382, 0.9239445975461231, 8.0
T-Test Confidence Interval = -6632.437535329434, 7224.437535329434
```

q9 was faster (6%) and q72 was slower (2.2%). I am going to focus more testing on query72 to see if this is caused perhaps by calling `unpack` or something else, will also test against a 23.04 baseline jar.

```
Name = query9
Means = 9498.6, 8937.0
Time diff = 561.6000000000004
Speedup = 1.0628398791540785
T-Test (test statistic, p value, df) = 2.5673438616859325, 0.03326472469818815, 8.0
T-Test Confidence Interval = 57.16740662534335, 1066.0325933746574
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed
--------------------------------------------------------------------
Name = query72
Means = 118548.6, 121185.4
Time diff = -2636.7999999999884
Speedup = 0.9782416033614612
T-Test (test statistic, p value, df) = -2.7090393198155214, 0.02669776978463036, 8.0
T-Test Confidence Interval = -4881.312162957456, -392.2878370425201
ALERT: significant change has been detected (p-value < 0.05)
ALERT: regression in performance has been observed
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -202,6 +209,194 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+      initialSpillPriority: Long,
+      spillCallback: SpillCallback,
+      needsSync: Boolean): RapidsBufferHandle = synchronized {
+    logWarning(s"Adding buffer ${id} to ${deviceStorage}")
```
logDebug

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+
+        while (!exhausted && !rmmShouldRetryAlloc &&
+            store.currentSpillableSize > targetTotalSize) {
+          val mySpillCount = spillCount.incrementAndGet()
```
- read here, don't increment.
- get lock
- re-read if I am the same one spill, else retry.

That way the first one will win, rather than the last one.

abellina replied:
Otherwise this could livelock. If the lock is held when incrementing, you guarantee that at least 1 thread has acquired the lock and could spill.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+      // logs where we say we spilled 0 bytes from X store
+      None
+    } else {
+      Some(0)
```
comment on what this means here.

abellina replied:
just return Some(totalSpilled) and put the rmm check as the if.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+      stream: Cuda.Stream): Unit = synchronized {
+    if (buffer.addReference()) {
+      withResource(buffer) { _ =>
+        logWarning(s"Spilling $buffer ${buffer.id} to ${spillStore.name}")
```
logDebug

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:
```diff
@@ -44,67 +39,60 @@ class RapidsHostMemoryStore(
-        val targetSize = math.max(currentSize - size, 0)
-        synchronousSpill(targetSize)
-      }
+    val allocation = addressAllocator.allocate(size)
```
try to get all of memory from address space pool and THEN go to `HostMemoryBuffer.allocate`, so this should move to ~ line 50, and we shouldn't need to check that size. We would instead blindly allocate if addressAllocator OOMs

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+  /**
+   * If `spillStore` defines a maximum size, make room for `buffer` by spilling.
+   */
+  private def makeRoomForBufferIfNeeded(
```
tryMakeRoom  because we have no guarantees that the memory is actually freed.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+    buffer: RapidsBuffer,
+    memoryBuffer: MemoryBuffer,
+    stream: Cuda.Stream): RapidsBuffer = synchronized {
+    val newBuffer = deviceStorage.copyBuffer(
```
make sure two threads calling here won't re-materialize the buffer into device memory.

abellina replied:
this needs to return an old RapidsBuffer if it already existed in the device tier.

abellina commented:
@jlowe @revans2 this is ready for another look. I have not tested unspill, but will do so shortly, will add a unit test around it.

abellina commented:
build

abellina commented:
build

abellina commented:
On the query72 performance, I re-ran this 10 times for my patch vs 23.04 baseline, and the results now show an improvement. Since before we showed a regression, I think this is just noise:

```
Name = query72
Means = 119955.1, 118587.8
Time diff = 1367.300000000003
Speedup = 1.0115298538298205
T-Test (test statistic, p value, df) = 2.455226321496253, 0.02447766025037065, 18.0
T-Test Confidence Interval = 197.30981193828916, 2537.2901880617164
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed
```

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  private def synchronousSpill(
```
Again, why are these separate? The private version that takes a Cuda.Stream is not called anywhere, except just above here. Also is there any time that we would not want to use the default stream?

abellina replied:
I can merge the methods, good eye.

On the default stream: I think default stream makes sense for PTDS, but for non-PTDS. We are not prepared to handle that optimally.

- The expectation is that buffers added to the spill store are synchronized w.r.t the stream that is doing the adding (`addBuffer` winds up synchronizing, unless it's from the shuffle code which guarantees it synchronizes buffers before calling `catalog.addBuffer`). 

- On spill, the PTDS stream is doing the copying, so it is the one responsible for synchronization, and default stream seems like the right choice for PTDS (but not for non-PTDS, this code is going to over synchronize on spill).

- On un-spill, or the memcpy-and-throwaway that happen if unspill=false, we are using the `cudaMemcpy` on the default stream (forced, the stream is not passed to those apis from calling code) and we are synchronizing on this default stream. Again, this is mostly OK because the result of the memcpy is after that the buffer is re-added to the store as a `RapidsDeviceMemoryBuffer` or returned to the user as a pointer to `DeviceMemoryBuffer`.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala:
```diff
@@ -50,24 +50,47 @@ abstract class RapidsBufferStore(
       (o1: RapidsBufferBase, o2: RapidsBufferBase) =>
         java.lang.Long.compare(o1.getSpillPriority, o2.getSpillPriority)
+    // buffers: contains all buffers in this store, whether spillable or not
     private[this] val buffers = new java.util.HashMap[RapidsBufferId, RapidsBufferBase]
```
This is being modified/read without the store's lock held.  What lock is supposed to be held when doing this?

abellina replied:
`buffers` is modified with the `BufferTracker` lock held. This line is happening during initialization, which should be happening before tasks start to run. So I think this is safe, but let me know if I've missed something.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala:
```diff
@@ -134,135 +213,74 @@ abstract class RapidsBufferStore(
+      if (memNeeded > 0 && memNeeded <= pendingFreeBytes.get) {
+        // This could be a futile wait if the thread(s) holding the pending buffers
+        // open are here waiting for more memory.
+        memoryFreedMonitor.wait(RapidsBufferStore.FREE_WAIT_TIMEOUT)
```
This is a bit confusing. The docs say that we will wait for `targetTotalSize` to be available before waking up, or timed out.  But this only waits for any buffer to be freed before waking up. Also the code is doing an if/not a while or anything like that. I think @jlowe wanted us to exit sooner, which is fine, but then we should update the docs to indicate what this is really doing.

abellina replied:
With this https://github.com/NVIDIA/spark-rapids/pull/7572/commits/bc2a2150bfa8be133a14e9ee5a465eb0fd9d10ab we don't have the waits anymore.

revans2 reviewed (commented):
I did a pass through the code, but I need to do a deep dive into the synchronization on it.

abellina commented:
For the stress testing I used query95 at 3/10/30 TB, and query72 at 3 TB with a reduced foot print. My main goal was to find deadlocks if anything, and cause spill. I am not seeing such issues.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+        while (!exhausted && !rmmShouldRetryAlloc &&
+            store.currentSpillableSize > targetTotalSize) {
+          val mySpillCount = spillCount
+          var spilled = false
```
I think I partially agree. The main difference is that with `totalSpilled > 0` we will not wait for pending bytes, because once the first buffer is spilled this total stays above 0. We also won't set `exhausted=true` and will rely on `store.currentSpillableSize` to be <= targeTotalSize to exit the loop. I don't think this is a live lock scenario, but it is different behavior, and I think would make us more aggressive at attempting to realloc.

I would then want to remove the `waitForPending` wait. It seems we would stop needing that.

jlowe replied:
Yeah, per our previous discussions, I think we would want to free what we can without waiting and retry the alloc rather than assume the alloc retry will fail if we freed something but it may not be enough and assume we need to wait for more memory to free before retrying.

abellina replied:
Ok I'll push a commit that cleans this up then, it should remove a question @revans2 raised on `waitForPending` as well

abellina replied:
https://github.com/NVIDIA/spark-rapids/pull/7572/commits/bc2a2150bfa8be133a14e9ee5a465eb0fd9d10ab This removes the waiting logic.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -202,6 +208,194 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+   * Adds a buffer to the device storage. This does NOT take ownership of the
+   * buffer, so it is the responsibility of the caller to close it.
+   *
+   * This version of `addBuffer` creates a `TempSpillBufferId` to use to refer to
```
Yeah, exactly. I'll just make the shuffle comment instead :)

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -225,7 +419,10 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+    acquireBuffer(handle.id)
+  }
+
+  private def acquireBuffer(id: RapidsBufferId): RapidsBuffer = {
```
Nice catch

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -202,6 +208,194 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+   *           brand new to the store, or the `RapidsBuffer` is invalid and
+   *           about to be removed).
+   */
+  def getExistingRapidsBufferAndAcquire(buffer: DeviceMemoryBuffer): Option[RapidsBuffer] = {
```
Yes it is to make it static. You can also mark it private in the object and then only a class with the same name can use it.

abellina replied:
https://github.com/NVIDIA/spark-rapids/pull/7572/commits/a0b039a04ddbe0a06e2d800ef2a74b2c56244da5 this commit moves the function to the object, and also marks it private.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+                val spillStore = store.spillStore
+                if (spillStore == null) {
+                  throw new OutOfMemoryError("Requested to spill without a spill store")
+                }
```
should be good now

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+
+  /**
+   * Given a specific `RapidsBuffer` spill it to `spillStore`
+   * @note called with catalog lock held
```
removed the synchronized statement

abellina commented:
build

abellina commented:
build

abellina commented:
build

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -304,8 +468,173 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
+    // total amount spilled in this invocation
+    var totalSpilled: Long = 0
+
+    if (store.currentSpillableSize > targetTotalSize) {
```
Not necessarily to fix for this PR, but one issue with this old logic is that it will not free _anything_ if targetTotalSize is larger than the current spillable size.  That seems pretty bad.  For example, freeing just 256MB of memory might allow a 1GB allocation to succeed.  Honestly not sure why we would care what the current amount of spillable memory is, we should just blindly ask the store to spill until we can allocate, and the store either will spill or it won't.  We shouldn't give up until the spill store is exhausted of spillable memory or we succeed in the allocation, whichever comes first.

abellina replied:
Filed this to track https://github.com/NVIDIA/spark-rapids/issues/7706

abellina commented:
@revans2 @jlowe I had introduced a bug in RapidsGdsStoreSuite. Fixed it with https://github.com/NVIDIA/spark-rapids/pull/7572/commits/6cca74d4a6b396b6aa730afe1f0c0b8612e3bc56.

abellina commented:
build
